File: python/knapsack/knapsack.py
class Knapsack:

    @staticmethod
    def solve(capacity, weights, values):

        n = len(values)

        table = []

        for _ in range(n + 1):
            row = [0] * (capacity + 1)
            table.append(row)

        for i in range(1, n + 1):

            for current_capacity in range(capacity + 1):

                if weights[i - 1] <= current_capacity:

                    include_item = (
                        values[i - 1]
                        + table[i - 1][
                            current_capacity - weights[i - 1]
                        ]
                    )

                    exclude_item = table[i - 1][current_capacity]

                    table[i][current_capacity] = max(
                        include_item,
                        exclude_item
                    )

                else:

                    table[i][current_capacity] = table[i - 1][
                        current_capacity
                    ]

        return table[n][capacity]

File: go/knapsack/knapsack.go
package knapsack

func Solve(
	capacity int,
	weights []int,
	values []int,
) int {

	n := len(values)

	table := make([][]int, n+1)

	for i := range table {
		table[i] = make([]int, capacity+1)
	}

	for i := 1; i <= n; i++ {

		for currentCapacity := 0; currentCapacity <= capacity; currentCapacity++ {

			if weights[i-1] <= currentCapacity {

				includeItem :=
					values[i-1] +
						table[i-1][currentCapacity-weights[i-1]]

				excludeItem :=
					table[i-1][currentCapacity]

				if includeItem > excludeItem {

					table[i][currentCapacity] =
						includeItem

				} else {

					table[i][currentCapacity] =
						excludeItem
				}

			} else {

				table[i][currentCapacity] =
					table[i-1][currentCapacity]
			}
		}
	}

	return table[n][capacity]
}
